In [1]:
!pip install PyMuPDF

In [2]:
import pymupdf
import re
import pandas as pd
from collections import Counter

# ref : https://www.slideshare.net/slideshow/a-collection-of-mutawatir-hadith/33190237

doc = pymupdf.open("C:\\Users\\Administrator\\Downloads\\Mutwatir Ahadith.pdf") 
all_texts = []
for page in doc: 
  text = page.get_text() 
  all_texts.append(text) 

with open("Final_Mutwatir_Ahadith.txt", "w", encoding="utf-8") as file:
    for hadith in all_texts:
        file.write(hadith + "\n\n") 

print("All_hadiths saved to Final_Mutwatir_Ahadith.txt")

All_hadiths saved to Final_Mutwatir_Ahadith.txt


In [3]:

with open("Final_Mutwatir_Ahadith.txt", "r", encoding="utf-8") as file:
    content = file.read()


count = content.count("Sahih Muslim [Arabic only]:")

print(f'The phrase "Sahih Muslim [Arabic only]:" appears {count} times.')


The phrase "Sahih Muslim [Arabic only]:" appears 320 times.


In [4]:

numbers = re.findall(r"Sahih Muslim \[Arabic only\]: (\d+)", content)
numbers = [int(num) for num in numbers]

print(numbers)

[2, 10, 29, 30, 33, 41, 47, 48, 50, 52, 57, 59, 72, 73, 83, 86, 98, 140, 163, 202, 220, 226, 228, 234, 238, 239, 245, 251, 267, 269, 293, 299, 317, 322, 323, 330, 331, 337, 346, 354, 356, 362, 367, 374, 401, 402, 404, 430, 432, 439, 460, 462, 486, 487, 488, 498, 569, 586, 588, 595, 599, 609, 614, 615, 617, 625, 777, 799, 802, 805, 810, 812, 822, 823, 824, 826, 828, 832, 852, 853, 854, 856, 857, 862, 870, 872, 874, 879, 922, 923, 925, 972, 976, 1002, 1006, 1007, 1034, 1038, 1064, 1082, 1099, 1105, 1112, 1114, 1118, 1119, 1122, 1123, 1215, 1219, 1223, 1228, 1233, 1261, 1274, 1285, 1288, 1322, 1332, 1355, 1367, 1368, 1371, 1393, 1395, 1396, 1397, 1398, 1401, 1402, 1403, 1490, 1498, 1499, 1509, 1516, 1518, 1521, 1522, 1536, 1543, 1549, 1578, 1586, 1676, 1677, 1687, 1737, 1739, 1761, 1762, 1771, 1776, 1795, 1808, 1835, 1838, 1851, 1863, 1897, 1901, 2513, 1909, 1910, 1912, 1914, 2008, 2009, 2108, 2131, 2159, 2161, 2164, 2168, 2177, 2194, 2201, 2246, 2412, 2413, 2449, 2463, 2465, 2466, 2467, 

In [5]:
len(numbers)

320

In [6]:
# Find all sequential numbers followed by a dot (e.g., 1., 2., 3., ...)
matches = re.findall(r"\b\d+\.", content)

# Count the occurrences
count = len(matches)

print(f'The count of sequential numbers followed by a dot is: {count}')

The count of sequential numbers followed by a dot is: 322


There are total of 322 hadiths but ocr could extract only 320 hadith numbers,lets try to find missing 2 hadiths


In [7]:

with open("Final_Mutwatir_Ahadith.txt", "r", encoding="utf-8") as file:
    content = file.read()

# Split content based on numbered paragraphs (e.g., "1. ", "2. ", "3. " etc.)
paragraphs = re.split(r"(?<=\n)\d+\.\s", content)

# Track paragraph numbers
invalid_paragraphs = []

for i, para in enumerate(paragraphs[1:], start=1):  # Skipping the first split (empty before 1.)
    number = re.findall(r"Sahih Muslim \[Arabic only\]: (\d+)", para)
    if len(number) < 1:
        print(f"................IDX....................... ",i)
        print(para)
        invalid_paragraphs.append(i)

# Print paragraphs that do not end with the phrase
print("Paragraphs that do not end with 'Sahih Muslim [Arabic only]:' are:", invalid_paragraphs)


................IDX.......................  91
Abu Hurairah, may Allah be pleased with him, reported: The Messenger of Allah 
(may peace be upon him) said: When anyone of you utters Tashahhud (in the 
Prayer) he must seek refuge with Allah from four (trials) and should thus say: O 
Allah! I seek refuge with You from the torment of Hell, from the torment of the 
grave, from the trial of life and death and from the evil of the trial of Messiah Al-
Dajjal(Antichrist). 
 

................IDX.......................  110
  I did not see the Messenger of Allah (may peace be upon him) reciting (the 
Qur'an) in the Night … 
'A'ishah, may Allah be pleased with her, reported: I did not see the Messenger of Allah (may 
peace be upon him) reciting (the Qur'an) in the Night Prayer in a sitting position, till he 
grew old and then he recited (it) in a sitting position, but when thirty or forty verses were 
left out of the surah, he would then stand up, recite them and bow down. 

Paragraphs that do 

Well as we can see that from the pdf file 91st and 110th hadith doesn't end with 'Sahih Muslim [Arabic only]:' phrase as reference

In [8]:
unique_numbers = set(numbers)  # Convert list to a set to remove duplicates

# Count the unique numbers
unique_count = len(unique_numbers)
print(f"Total numbers: {len(numbers)}")
print(f"Unique numbers: {unique_count}")


Total numbers: 320
Unique numbers: 319


let's find the duplicate hadith

In [9]:

number_counts = Counter(numbers)

duplicates = {num: count for num, count in number_counts.items() if count > 1}

print("Duplicate numbers and their occurrences:")
for num, count in duplicates.items():
    print(f"Number {num} appears {count} times")


Duplicate numbers and their occurrences:
Number 5245 appears 2 times


first i thought it's mistake of OCR but after manually checking the pdf file i can confirm that 320 and 322 number hadith refers to same 5245 number hadith from sahih muslim book

In [10]:
# collected the data from : https://www.kaggle.com/datasets/abdokamr/arabic-hadith
muslim = pd.read_csv("Sahih Muslim.csv")
muslim['Sahih Muslim'][5]

'و حَدَّثَنَا عُبَيْدُ اللَّهِ بْنُ مُعَاذٍ الْعَنْبَرِيُّ حَدَّثَنَا أَبِي ح و حَدَّثَنَا مُحَمَّدُ بْنُ الْمُثَنَّى حَدَّثَنَا عَبْدُ الرَّحْمَنِ بْنُ مَهْدِيٍّ قَالَا حَدَّثَنَا شُعْبَةُ عَنْ خُبَيْبِ بْنِ عَبْدِ الرَّحْمَنِ عَنْ حَفْصِ بْنِ عَاصِمٍ قَالَ قَالَ رَسُولُ اللَّهِ صَلَّى اللَّهُ عَلَيْهِ وَسَلَّمَ كَفَى بِالْمَرْءِ كَذِبًا أَنْ يُحَدِّثَ بِكُلِّ مَا سَمِعَ و حَدَّثَنَا أَبُو بَكْرِ بْنُ أَبِي شَيْبَةَ حَدَّثَنَا عَلِيُّ بْنُ حَفْصٍ حَدَّثَنَا شُعْبَةُ عَنْ خُبَيْبِ بْنِ عَبْدِ الرَّحْمَنِ عَنْ حَفْصِ بْنِ عَاصِمٍ عَنْ أَبِي هُرَيْرَةَ عَنْ النَّبِيِّ صَلَّى اللَّهُ عَلَيْهِ وَسَلَّمَ بِمِثْلِ ذَلِكَ.'

Make Final_Mutwatir_Ahadith.csv file for all 322 hadiths

In [11]:
hadith_data = []
for num in numbers:
    hadith = muslim['Sahih Muslim'][num-1]
    hadith_data.append([num, hadith])

hadith_df = pd.DataFrame(hadith_data, columns=["hadith number", "hadith"])
hadith_df.to_csv("Mutawatir_sahih_muslim.csv", index=False)
hadith_df

hadith number                                             hadith
0                2  و حَدَّثَنَا أَبُو بَكْرِ بْنُ أَبِي شَيْبَةَ ...
1               10  و حَدَّثَنَا أَبُو بَكْرِ بْنُ أَبِي شَيْبَةَ ...
2               29  حَدَّثَنَا قُتَيْبَةُ بْنُ سَعِيدٍ حَدَّثَنَا ...
3               30  و حَدَّثَنَا أَبُو الطَّاهِرِ وَحَرْمَلَةُ بْن...
4               33  حَدَّثَنَا أَبُو غَسَّانَ الْمِسْمَعِىُّ مَالِ...
..             ...                                                ...
315           5231  حَدَّثَنَا شِهَابُ بْنُ عَبَّادٍ الْعَبْدِيُّ ...
316           5236  حَدَّثَنِي عَلِيُّ بْنُ حُجْرٍ السَّعْدِيُّ حَ...
317           5245  حَدَّثَنَا مُحَمَّدُ بْنُ الْمُثَنَّى وَمُحَمّ...
318           5360  حَدَّثَنَا أَبُو بَكْرِ بْنُ أَبِي شَيْبَةَ حَ...
319           5245  حَدَّثَنَا مُحَمَّدُ بْنُ الْمُثَنَّى وَمُحَمّ...

[320 rows x 2 columns]

In [12]:
hadith_df.hadith[1]

'و حَدَّثَنَا أَبُو بَكْرِ بْنُ أَبِي شَيْبَةَ وَزُهَيْرُ بْنُ حَرْبٍ جَمِيعًا عَنْ ابْنِ عُلَيَّةَ قَالَ زُهَيْرٌ حَدَّثَنَا إِسْمَعِيلُ بْنُ إِبْرَاهِيمَ عَنْ أَبِي حَيَّانَ عَنْ أَبِي زُرْعَةَ بْنِ عَمْرِو بْنِ جَرِيرٍ عَنْ أَبِي هُرَيْرَةَ قَالَ كَانَ رَسُولُ اللَّهِ صَلَّى اللَّهُ عَلَيْهِ وَسَلَّمَ يَوْمًا بَارِزًا لِلنَّاسِ فَأَتَاهُ رَجُلٌ فَقَالَ يَا رَسُولَ اللَّهِ مَا الْإِيمَانُ قَالَ أَنْ تُؤْمِنَ بِاللَّهِ وَمَلَائِكَتِهِ وَكِتَابِهِ وَلِقَائِهِ وَرُسُلِهِ وَتُؤْمِنَ بِالْبَعْثِ الْآخِرِ قَالَ يَا رَسُولَ اللَّهِ مَا الْإِسْلَامُ قَالَ الْإِسْلَامُ أَنْ تَعْبُدَ اللَّهَ وَلَا تُشْرِكَ بِهِ شَيْئًا وَتُقِيمَ الصَّلَاةَ الْمَكْتُوبَةَ وَتُؤَدِّيَ الزَّكَاةَ الْمَفْرُوضَةَ وَتَصُومَ رَمَضَانَ قَالَ يَا رَسُولَ اللَّهِ مَا الْإِحْسَانُ قَالَ أَنْ تَعْبُدَ اللَّهَ كَأَنَّكَ تَرَاهُ فَإِنَّكَ إِنْ لَا تَرَاهُ فَإِنَّهُ يَرَاكَ قَالَ يَا رَسُولَ اللَّهِ مَتَى السَّاعَةُ قَالَ مَا الْمَسْئُولُ عَنْهَا بِأَعْلَمَ مِنْ السَّائِلِ وَلَكِنْ سَأُحَدِّثُكَ عَنْ أَشْرَاطِهَا إِذَا وَلَد